src/wallet-redux.js

```diff
@@ -22,137 +24,246 @@ export const WalletType = {
+            // Remove wallet
+            const wallets = new Map(state.wallets);
+            wallets.delete(action.walletId);
+
+            // Remove wallet's accounts (including contracts)
+            const accounts = new Map(state.accounts);
+            for (const account of [...state.accounts.values()]) {
+                if (account.walletId === action.walletId) {
+                    accounts.delete(account);
+                }
+            }
 
-            const newEntries = action.wallets.map(x => {
-                const oldEntry = state.entries.get(x.id);
+            return updateState({
+                wallets,
+                accounts,
+            });
+        }
 
-                return [
-                    x.id,
-                    Object.assign({}, oldEntry, x)
-                ];
+        case TypeKeys.POPULATE: {
+            const wallets = new Map(state.wallets);
+            const accounts = new Map(state.accounts);
+
+            action.listedWallets.forEach(wallet => {
+                if (wallet.type !== WalletType.LEGACY) {
+                    const entry = {
+                        id: wallet.accountId,
+                        label: wallet.label,
+                        type: wallet.type,
+                        fileExported: wallet.fileExported,
+                        wordsExported: wallet.wordsExported,
+                    };
+
+                    // merge with previous information
+                    wallets.set(
+                        wallet.accountId,
+                        Object.assign({}, wallets.get(wallet.accountId), entry)
+                    );
+                }
+
+                wallet.addresses.forEach(address => {
+                    const entry = {
+                        address: address.address,
+                        label: address.label,
+                        type: wallet.type === WalletType.LEDGER ? AccountType.LEDGER : AccountType.KEYGUARD_HIGH,
+                        isLegacy: wallet.type === WalletType.LEGACY,
+                        walletId: wallet.accountId,
+                    };
+
+                    // merge with previous information
+                    accounts.set(
+                        address.address,
+                        Object.assign({}, accounts.get(address.address), entry)
+                    );
+                });
+
+                wallet.contracts.forEach(contract => {
+                    const entry = Object.assign({}, contract, {
+                        type: contract.type === 1 /* Nimiq.Account.Type.VESTING */ ? AccountType.VESTING : AccountType.HTLC,
+                        stepAmount: contract.stepAmount / 1e5,
+                        totalAmount: contract.totalAmount / 1e5,
+                        label: contract.label,
+                        owner: contract.owner, // vesting only
+                        sender: contract.sender, // HTLC only
+                        recipient: contract.recipient, // HTLC only
```

Comment:
Ups, I just noticed the addition of `label, owner, sender, recipient` is unnecessary, because here we are _extending_ the `contract` object with `Object.assign()`. That's why those props were not in the original code. I'm stupid. You can remove them again.

Reply:
This code is only overwriting the props which need to be adapted for the Safe from the AM values (such as Luna to NIM) and adding necessary new props.

Reply:
Oh no, and I fell for you ;(